syntax = "proto3";

package precompiled;

message Function {
  bytes byte_code = 1;
}

message QualifiedType {
  int64 type_id = 1;
  int32 qualifiers = 2;
}

message FunctionTypeDefinition {
  message Parameter {
    string name = 1;
    int64 type = 2;
    uint32 flags = 3;
  }
  repeated Parameter parameter = 1;
  repeated int32 return_type = 2;
  bool eager = 3;
}

message TypeDefinition {
  oneof type {
    int64 primitive = 1;
    int64 pointer = 2;
    int64 buffer_pointer = 3;
    FunctionTypeDefinition function = 4;
    int64 slice = 5;
  }
}

message TypeSystem {
  repeated TypeDefinition type = 1;
}
